patches/server/1031-Add-experience-points-API.patch

```diff
@@ -0,0 +1,60 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Lukas Planz <lukas.planz@web.de>
+Date: Tue, 5 Sep 2023 20:34:20 +0200
+Subject: [PATCH] Add experience points API
+
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+index 4e6fea7cf11b1e29ae7c7098a6f5d06bb5f93cc2..c1274b96c22ff183341369df374483bea791d543 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -1802,6 +1802,49 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         Preconditions.checkArgument(exp >= 0, "Total experience points must not be negative (%s)", exp);
+         this.getHandle().totalExperience = exp;
+     }
++    // Paper start
++    @Override
++    public int getTotalExperiencePoints() {
++        return getTotalExperiencePoints(this.getLevel()) + Math.round(this.getExperiencePointsNeededForNextLevel() * getExp());
++    }
++
++    @Override
++    public void setTotalExperiencePoints(final int exp) {
++        Preconditions.checkArgument(exp >= 0, "Total experience points must not be negative (%s)", exp);
++        int level = getLevelsForExperiencePoints(exp);
++        int remainingPoints = exp - getTotalExperiencePoints(level);
++
++        this.getHandle().experienceLevel = level;
++        this.getHandle().experienceProgress = (float) remainingPoints / this.getExperiencePointsNeededForNextLevel();
++        this.getHandle().lastSentExp = -1;
++    }
++
++    @Override
++    public int getExperiencePointsNeededForNextLevel() {
++        return this.getHandle().getXpNeededForNextLevel();
++    }
++
++    // See https://minecraft.fandom.com/wiki/Experience#Leveling_up for reference
++    private int getTotalExperiencePoints(int level) {
++        if (level <= 16) {
++            return (int) (Math.pow(level, 2) + 6 * level);
++        } else if (level <= 31) {
++            return (int) (2.5 * Math.pow(level, 2) - 40.5 * level + 360.0);
++        } else {
++            return (int) (4.5 * Math.pow(level, 2) - 162.5 * level + 2220.0);
++        }
++    }
++
++    private int getLevelsForExperiencePoints(int points) {
++        if (points <= 352) { // Level 0-16
++            return (int) Math.floor(Math.sqrt(points + 9) - 3);
++        } else if (points <= 1507) { // Level 17-31
++            return (int) Math.floor(8.1 + Math.sqrt(0.4 * (points - (7839.0 / 40.0))));
++        } else { // 32+
++            return (int) Math.floor((325.0 / 18.0) + Math.sqrt((2.0 / 9.0) * (points - (54215.0 / 72.0))));
++        }
++    }
++    // Paper end
```

Comment:
I'm not a super fan of copying the logic for converting level into total points. Although I'm not super clear on what a good way to refactor the vanilla logic would be to allow us to use it directly. I think at the minimum, some `// Paper - diff on change` comments should be added to the key lines in `Player#getXpNeededForNextLevel` as that seems to be where the logic comes from.

Reply:
Yeah, I agree that this is a somewhat unfortunate way of handling this, but due to how the experience stuff is implemented it is necessary.

If you look at the initial commit of this PR, you can see that at least for setting I was able to use parts of vanilla code to set the level and progress. Getting still had to be done with the formula.

Reply:
Yeah, then at least add some `// Paper - diff on change` comments in a few places to try and catch if mojang ever changes this.